q2_moshpit/kraken2/classification.py

```diff
@@ -47,15 +51,92 @@ def _construct_output_paths(
+        quick=False,
+        report_minimizer_data=False
+):
+    kwargs = {k: v for k, v in locals().items()
+              if k not in ["seqs", "kraken2_db", "ctx"]}
+
+    _classify_kraken2 = ctx.get_action("moshpit", "_classify_kraken2")
+    collate_kraken2_reports = ctx.get_action("moshpit",
+                                             "collate_kraken2_reports")
+    collate_kraken2_outputs = ctx.get_action("moshpit",
+                                             "collate_kraken2_outputs")
+
+    if seqs.type <= SampleData[SequencesWithQuality]:
+        partition_method = ctx.get_action("demux", "partition_samples_single")
+    elif seqs.type <= SampleData[PairedEndSequencesWithQuality]:
+        partition_method = ctx.get_action("demux", "partition_samples_paired")
+    # FeatureData[MAG] is not parallelized
+    elif seqs.type <= FeatureData[MAG]:
+        kraken2_reports, kraken2_outputs = \
+                _classify_kraken2(seqs, kraken2_db, **kwargs)
+        return kraken2_reports, kraken2_outputs
+    else:
+        raise NotImplementedError()
+
+    (partitioned_seqs,) = partition_method(seqs)
+
+    kraken2_reports = []
+    kraken2_outputs = []
+    for seq in partitioned_seqs.values():
+        (kraken2_report, kraken2_output) = _classify_kraken2(
+                seq, kraken2_db, **kwargs)
+        kraken2_reports.append(kraken2_report)
+        kraken2_outputs.append(kraken2_output)
+
+    (collated_kraken2_reports,) = collate_kraken2_reports(kraken2_reports)
+    (collated_kraken2_outputs,) = collate_kraken2_outputs(kraken2_outputs)
+
+    return collated_kraken2_reports, collated_kraken2_outputs
+
+
 def _classify_kraken2(
+        seqs: Union[
+            SingleLanePerSamplePairedEndFastqDirFmt,
+            SingleLanePerSampleSingleEndFastqDirFmt,
+            MAGSequencesDirFmt,
+        ],
+        kraken2_db: Kraken2DBDirectoryFormat,
+        threads: int = 1,
+        confidence: float = 0.0,
+        minimum_base_quality: int = 0,
+        memory_mapping: bool = False,
+        minimum_hit_groups: int = 2,
+        quick: bool = False,
+        report_minimizer_data: bool = False
+) -> (
+        Kraken2ReportDirectoryFormat,
+        Kraken2OutputDirectoryFormat,
+):
+    kwargs = {k: v for k, v in locals().items()
+              if k not in ["seqs", "kraken2_db", "ctx"]}
+
+    common_args = _process_common_input_params(
+        processing_func=_process_kraken2_arg, params=kwargs
+    )
+    common_args.extend(["--db", str(kraken2_db.path)])
+    return classify_kraken2_helper(seqs, common_args)
+
+
+def classify_kraken2_helper(
         seqs, common_args
 ) -> (Kraken2ReportDirectoryFormat, Kraken2OutputDirectoryFormat):
     if isinstance(seqs, MAGSequencesDirFmt):
         manifest = None
     else:
         manifest: Optional[pd.DataFrame] = seqs.manifest.view(pd.DataFrame)
 
-    base_cmd = ["kraken2", *common_args]
+    base_cmd = ["kraken2", *common_args, '--memory-mapping']
```

Comment:
This `--memory-mapping` flag prevents the database from being loaded into RAM for every thread. It slows things down compared to loading the database for every thread, but it prevents us from needing to load a ~60GB database into RAM a million times. For now we'll have it here to be safe, but we can definitely be smarter about when we do and do not need it and how we handle the database (/dev/shm for instance).

Reply:
I see the action's current default is False (no memory mapping) - do you think it should stay that way with the addition of parallel support? I could see arguments either way.